### ISSUE TITLE
Add filter capability for exact match or fields without data

### ISSUE THREAD

**Issue by RDmitchell** [SEED-platform/seed#20]
FILTERING: Can’t specify exact match
When doing the filtering, can't specify an "exact match" -- so there is not a way to see only records that have Use Description set to "Multi-Family", not including "Multi-Family Condominium". Could specify exact matches as surrounded in "quotes" (?) or some other scheme. This might enable blank fields to be filtered by “”

(LBNL Ticket # 124)

**Comment by RDmitchell** [comment]
See also Issue # 64, which I'm not sure I completely understand. It is related but not the same. We may have to talk to Marshall (DC) when we get to this issue.

**Comment by RDmitchell** [comment]
Megha -- I am assigning this issue to you to investigate whether this is something that you can do through the UI or whether it is "back end" work.

**Comment by RDmitchell** [comment]
Here is a screen shot that shows the issue. In the Building List, you can see that some records have no data in the Energy Star Score field. The users would like to know which records have no data in this field, but right now there is now way to filter on "null" data. You can sort by a field and have the blanks turn up and the beginning or the end, but that is not an efficient way for users to deal with these records when they may have hundreds of records with this issue. 

One way to do this would be to have the user enter two double quotes together (?) "" if they want the program to return records where the fields with this "" filter have null data. There is probably a nicer UI option, but this could be a quick fix for now, with the idea of making it more user friendly later.

![image](https://cloud.githubusercontent.com/assets/6314950/7169769/4d725f30-e37d-11e4-8631-050dac549ff4.png)

Then there is the issue of being able to filter on an exact match. Right now when you type something into the filtering field at the top of the header, the program looks for records that contain what was typed. So for example, if you wanted to filter on the Premises Occupancy Classification field, and just return "Commercial Condo", the program returns records that contain both just Commercial and just Condo, as well as those that contain Commercial Condo. So we need a way for users to specify that they want to search for an exact match on Commercial Condo.

Seems like one simple approach would be to do an exact match if the search term was enclosed in double quotes? So typing in "Commercial Condo" in the filter field would return an exact match on the combination of words. Again, we could add a more user friendly UI option later if we thought that was needed, but this would be potentially a quick way to get the functionality into the program.

![image](https://cloud.githubusercontent.com/assets/6314950/7169918/c36423f8-e37e-11e4-95a2-7a240ce3834d.png)

**Comment by nicholasserra** [comment]
Is there a preference on how we search for blank data? I'm wondering if the user experience will be weird if we allow them to hit enter on a blank field. I think an icon was suggested, but that might be odd too.

**Comment by RDmitchell** [comment]
My quick fix for this was for the user to enter "" for blank data, and
"text" for exact match.

Not super user friendly, but if it works, it will be great improvement. We
can get fancier later if we need to.

On Wed, Apr 22, 2015 at 11:14 AM, Nicholas Serra <notifications@github.com>
wrote:

> Is there a preference on how we search for blank data? I'm wondering if
> the user experience will be weird if we allow them to hit enter on a blank
> field. I think an icon was suggested, but that might be odd too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-95289099>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by nicholasserra** [comment]
I actually really like that. It would save having to do any FE work on this.

**Comment by RDmitchell** [comment]
yes, my thoughts exactly.

On Wed, Apr 22, 2015 at 11:24 AM, Nicholas Serra <notifications@github.com>
wrote:

> I actually really like that. It would save having to do any FE work on
> this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-95291779>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by RDmitchell** [comment]
Nick -- this is only going to work on fields with characters, not on numeric fields, is that correct? 

Also, did you make any unit tests or test files for this? If not, you might want to add a unit test.

**Comment by nicholasserra** [comment]
Should work on any fields, as it's just doing an `exact` or `null` search. Unit tests were included.

**Comment by RDmitchell** [comment]
I am testing on our seedtest environment (I can make you an account if you want to test there), and I am not seeing the behavior I thought I would. Some of what I am experiencing may not be from your changes. 

I have made a sample data set and it is in the following directory -- there are two files (a building list and a Portfolio Manager file, both generated from the standard sample data of 512 records, but with 19 records in each. I changed the address field in both so there should be an exact match across the two data sets on address)

https://drive.google.com/a/lbl.gov/folderview?id=0B3fTKpZ9Dx7LYmRNenF6TTJweTQ&usp=sharing

Using that data, importing just the Building list, and filtering in the Building list on not an exact match but just condo, i get back 4 buildings (after the program goes through some unexplained gyrations).
![image](https://cloud.githubusercontent.com/assets/6314950/7399442/eb62e532-ee6a-11e4-9264-30f271d3c6f9.png)

So then if I do an exact match, (not clearing the filter by deleting the first one, but just by retyping the new filter) entering "CONDO" into the filter field at the top of the column, I should get 2 buildings back. But the program seems like it is returning 0 buildings. Maybe I am not understanding how to enter the filter?

![image](https://cloud.githubusercontent.com/assets/6314950/7399512/598b60e8-ee6b-11e4-9464-07e937ebc2f1.png)

Then just to be sure I am starting from scratch, I clear the filter completely, so nothing is entered in the filter input box, and the program returns 19 buildings:

![image](https://cloud.githubusercontent.com/assets/6314950/7399568/9dd58828-ee6b-11e4-87b0-31d130f2c8f9.png)

Then I type in "CONDO" into the Premises Occupancy Classification (this is how it got mapped in my data, it will probably be mapped to something else in your data), remove focus from that filter input box (should I have to do that? I think maybe I do) and I still get 0 records returned. 

![image](https://cloud.githubusercontent.com/assets/6314950/7399620/deba135e-ee6b-11e4-89a8-27006cd612b7.png)

I am happy to jump on a webinar with you if you want to see this in action -- maybe I am doing something wrong, maybe there is something about our test environment that is set up differently than your development environment (?).

thx

**Comment by nicholasserra** [comment]
I'd love an account so I can test this out myself :)

**Comment by RDmitchell** [comment]
No problem -- I will add you right now and you will get an email to set the
password.

On Wed, Apr 29, 2015 at 12:42 PM, Nicholas Serra <notifications@github.com>
wrote:

> I'd love an account so I can test this out myself :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-97557367>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by RDmitchell** [comment]
I just made an org called Quick Left and added you as the owner. You will get an email to set the password, then you can log in and test away.

**Comment by nicholasserra** [comment]
What's the easiest way to get buildings into the building list? Locally I added them manually. I've uploaded both sample files and don't see any of those buildings appear in my list.

**Comment by RDmitchell** [comment]
You might want to watch the video in the seed wiki, under the Code Review
topic:

https://github.com/SEED-platform/seed/wiki/Code-Review

Under the Webinar Recordings section, the first one, UI Program Demo, is me
walking through the program from start to finish, including loading data.
For some reason, when you click on the link you have to enter a name (I
guess because it is access the video through our Readytalk account), but
then you will be able to get to the video.

And I am happy to walk you through any of this via webinar if needed, just
let me know.

On Wed, Apr 29, 2015 at 12:55 PM, Nicholas Serra <notifications@github.com>
wrote:

> What's the easiest way to get buildings into the building list? Locally I
> added them manually. I've uploaded both sample files and don't see any of
> those buildings appear in my list.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-97561999>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by nicholasserra** [comment]
I've watched it, but I'll watch it again. I guess I thought importing both data files would load those buildings into that view.

**Comment by RDmitchell** [comment]
Nick -- I actually haven't tested yet whether the building match. You can just test on one file, the building file. If you import that file into a dataset, map the fields, and save the mapping, the program should put those 19 records into the Building list. Don't worry about loading up the 2nd Portfolio Manager file.

**Comment by nicholasserra** [comment]
Oh. My fix was directly for the building view. Not sure if that specificity made a difference on whether it works for other pages. I think it might have, since I was only testing on the building view. Maybe this works for buildings only and just needs expanded.

**Comment by nicholasserra** [comment]
Got buildings into my list, and tested searching for `condo` without quotes. Got both `condo` and `commercial condo` back. Added quotes, and only got `condo` back, as expected. I think we need to do a little demo to figure this one out, as I'm not experiencing the issue you're seeing.

**Comment by RDmitchell** [comment]
Yes, sounds like we need a webinar. 

 

I can do it at 3 my time (Pacific) – does that work for you? I will send you call in info in a separate email

 

Robin 

***

Robin Mitchell

Building Technology and Energy Systems

Lawrence Berkeley National Laboratory

**Comment by nicholasserra** [comment]
That should be okay. Can we try a google Hangout, that way we can both show our screens? Readytalk has not been working out for me. No big deal if not.

**Comment by RDmitchell** [comment]
Rich Brown and I tested this functionality on seedtest today using his Mac and it didn't work for us. So it is a bit of a mystery why it works for you on seedtest (which it clearly does because the screenshots you sent today were using seedtest) and doesn't for us. This might be something that Miles, who can read the server logs, might have to help debug (??).

**Comment by RDmitchell** [comment]
If possible, do you think this could get straightened out early next week (say 5/4 or 5/5)? Thx

**Comment by nicholasserra** [comment]
Yes, looking into it today

**Comment by cycustodio** [comment]
I did the following tests using the covered-buildings file from Issue 20 without changing the mapping fields.
* In the filter for "Use Description", I tried different permutations for condo keyword and found how many buildings are returned.

Search Keyword | number of buildings
----------------------|-------------------------
CONDO| 4
"CONDO"| 2
"" | 6
condo | 4
"condo" | 0
CoNdO | 4
"CoNdO" | 0

When I try filtering Premises City, the null search is not working. It should return 4 buildings when searching for null values.
![premisescity_null](https://cloud.githubusercontent.com/assets/8461331/7464501/54f7f10c-f278-11e4-94dd-46b7afd66a5e.png)
![premisescity_energytown](https://cloud.githubusercontent.com/assets/8461331/7464508/6cfe2096-f278-11e4-9024-054f7753291e.png) 

Other fields where I tried the null search and it worked: Owner and State Province
![owner_null](https://cloud.githubusercontent.com/assets/8461331/7464645/6b7f6094-f279-11e4-8110-cd0ee5dd077b.png)
![stateprovince_null](https://cloud.githubusercontent.com/assets/8461331/7464646/6b8d0dfc-f279-11e4-9ae2-a08db250857e.png)

**Comment by nicholasserra** [comment]
Can you provide screenshots of both 'condo' (without quotes) and '"CONDO"' (with quotes and all caps)? It's hard to tell if exact matches was working based off of numbers. Sounds like a yes. 

via Hiptop

**Comment by cycustodio** [comment]
It works. 
condo (small caps, no quotes)
![usedescription_condo](https://cloud.githubusercontent.com/assets/8461331/7465035/61c579c6-f27e-11e4-9548-2533679cc997.png)

"CONDO" (all caps, with quotes)
![usedescription_condoquotes](https://cloud.githubusercontent.com/assets/8461331/7465036/61ebbdc0-f27e-11e4-8857-225417cc63db.png)

**Comment by nicholasserra** [comment]
Awesome thank you. Maybe there is something special about that first `Premises City` field that you tried the null value search on. Good to know what the exact and null searches are for the most part working.

**Comment by rebrownlbnl** [comment]
Could it be that "full-fledged" db fields are handled differently than
fields from the json blob? The filter makes an API call back to the db,
right?

**Comment by RDmitchell** [comment]
I think Use Description might be the unmapped field as it comes directly
from the input file.

Claudine, try the same test except during the mapping step change Use
Description to another name like Occupancy Classification and see if the
filter still works.

**Comment by cycustodio** [comment]
1) I changed the mapping for Property Type to Occupancy Type, and a further
check changed City to Use Description.

original file      :: old mapping       :: new mapping
Property Type :: Use Description :: Occupancy Type
City                 :: Premises City    :: Use Description
[image: Inline image 1]

Occupancy Type doesn't do the null search nor the exact search.
[image: Inline image 2]

Use Description (formerly Premises City) can do null and exact searches.
[image: Inline image 3]

[image: Inline image 4]


2) Might add this as a new issue,  but it looks like the filters are
getting saved and buildings with null values are disappearing. I uploaded
19 buildings, when I filter by State Province and remove the filter I
should get back all 19 buildings but I'm getting 16 buildings (3 buildings
have null values in this field). I checked the API request and its saving
previous filter parameters, at least the previous field names.

{"q":"",
"number_per_page":25,
"page":1,
"order_by":
"tax_lot_id",
"sort_reverse":true,
*"filter_params":{"Occupancy Type":"","state_province":""}*}



----------------------------------------------------------------
Claudine Y. Custodio
cycustodio@lbl.gov
510-495-8062 (Office)
Energy Technologies Area
Lawrence Berkeley National Lab
-----------------------------------------------------------------

**Comment by RDmitchell** [comment]
Claudine

Can you check to see if Use Description is a field in the building snapshot
table? It would be good to try to figure out when the filtering works and
when it doesn't possibly based on where the field is stored, how it is
mapped (?), etc.

It sounds like we might need to do a deep dive into how filtering works in
order to get this new feature to work consustently

Sigh ... i keep thinking these things are simple fixes but they never seem
to turn out that way.

**Comment by cycustodio** [comment]
I can confirm that use_description is part of the building snapshot table
since it's outside of the extra_data json blob.


----------------------------------------------------------------
Claudine Y. Custodio
cycustodio@lbl.gov
510-495-8062 (Office)
Energy Technologies Area
Lawrence Berkeley National Lab
-----------------------------------------------------------------

**Comment by RDmitchell** [comment]
So i suspect the filtering only works on fields that are in the building
snapshot table not including the json extra data blob field.

Can you do some more testing on both types of fields to confirm that?

And that would probably be true of any filtering not just this new feature.

**Comment by cycustodio** [comment]
Changed the mapping for the same file. 

original file | new mapping
--------------- | ---------------
City | Premises City
Property Type | Occupancy Type
AYB_YearBuilt | AYB_YearBuilt

The new mappings are part of the extra_data json. All three don't have exact and null filtering. 
![extradata_filter](https://cloud.githubusercontent.com/assets/8461331/7482899/657a26d2-f32f-11e4-818b-0b13ca6d16f6.png)

**Comment by nicholasserra** [comment]
Interesting. The intricacies are hard to follow. Can you give me steps to reproduce from a fresh account, or fresh data set? If I can reproduce locally I should be able to figure out a fix for this scenario.

**Comment by cycustodio** [comment]
I uploaded covered-buildings-sample.csv from this google folder.
https://drive.google.com/a/lbl.gov/folderview?id=0B3fTKpZ9Dx7LYmRNenF6TTJweTQ&usp=sharing

These are the field names used in Data Mapping. Then I saved the mapping and building matches. 
![extradata_mapping](https://cloud.githubusercontent.com/assets/8461331/7483187/be4a6ab8-f331-11e4-8b1d-ec1bf31311d6.png)

You should have 19 buildings in your Building list.

**Comment by nicholasserra** [comment]
Thank you. I'll give that a shot and see if I can reproduce.

**Comment by RDmitchell** [comment]
Nick

I think when you have been importing data you have not changed the field
names in the Mapping screen, which is why the filtering is working for you.
If you type in the new names Claudine shows above (on the left hand side of
the mapping screen) for the three fields ( on the right hand side of the
mapping screen ) you should be able to reproduce the problem. Let us know
if this doesn't  make sense and i will do a screen shot.

**Comment by nicholasserra** [comment]
I'm understanding the issue now. It's basically not working for non standard field names.

**Comment by nicholasserra** [comment]
So it looks like, if you do the mapping this way, it uses a completely different endpoint to do field searches. Gonna have to implement this on that search endpoint too. Basically, it's like expanding this feature to another view.

**Comment by nicholasserra** [comment]
Er,  misspoke. I was looking at searching while on the initial mapping page, before you click "Save mappings". We might want to expand this feature to that at some point.

Looks like for this issue specifically, I just need to hack into the json field filtering.

**Comment by RDmitchell** [comment]
when you say non-standard, what do you mean? There is almost always going
to be a different "mapping" of fields from the original data to the SEED
field names. So what do you consider to be "non-standard"?

On Mon, May 11, 2015 at 11:10 AM, Nicholas Serra <notifications@github.com>
wrote:

> I'm understanding the issue now. It's basically not working for non
> standard field names.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-101002496>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by nicholasserra** [comment]
Non standard meaning it is not a field in the database table, but instead a field in the JSON mapping.

**Comment by RDmitchell** [comment]
A field in the JSON mapping meaning a field in the extra_data JSON blob
field in the database?

On Mon, May 11, 2015 at 12:49 PM, Nicholas Serra <notifications@github.com>
wrote:

> Non standard meaning it is not a field in the database table, but instead
> a field in the JSON mapping.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SEED-platform/seed/issues/20#issuecomment-101029580>.
>



-- 
Robin Mitchell
Building Technology and Urban Systems Department
Environmental Energy Technologies Division
Lawrence Berkeley National Laboratory

**Comment by nicholasserra** [comment]
Yes. I have it all figured out, fix should be up today.

**Comment by RDmitchell** [comment]
I have tested this and it is working for any field that doesn't have the Min / Max range filter input. The response of the software to the filter request is pretty dog slow, but that's for another issue !

I am closing this issue.

**Comment by cycustodio** [comment]
File: Issue 20 covered buildings
Total Buildings: 19
"Property Type" is a non-standard mapping field, "City" is a standard mapping field. 

Filter by "Property Type": 13
Missing the buildings with null values for "Property Type". 

Property Type  |  Building Count
--------------|----------------
CONDO | 4
"CONDO" | 2
"" | 6
condo | 4
"condo" | 0
CoNdO | 4
"CoNdO" | 0

Filter by "City": 19

City | Building Count 
--------------|----------------
EnergyTown | 15
"EnergyTown" | 15
ENERGYTOWN | 15
"ENERGYTOWN" | 0
energytown | 15
"" | 4
